File: apps/web/src/lib/trusted_contacts.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';

import {
	MAX_TRUSTED_CONTACTS,
	normaliseTrustedContact,
	normaliseTrustedContacts,
	hasReachableChannel,
	type TrustedContact,
} from './trusted_contacts';

// ── normaliseTrustedContact ────────────────────────────────

test('normaliseTrustedContact: keeps trimmed name + populated optionals', () => {
	const out = normaliseTrustedContact({
		name: '  Alice  ',
		phone: ' +44 7700 900111 ',
		email: '  alice@example.com  ',
		relationship: ' partner ',
	});
	assert.deepStrictEqual(out, {
		name: 'Alice',
		phone: '+44 7700 900111',
		email: 'alice@example.com',
		relationship: 'partner',
	});
});

test('normaliseTrustedContact: drops empty-string optionals (does not store "")', () => {
	// A regression here would persist `{ phone: '' }` into the bag,
	// which downstream code would then have to defensively check for —
	// better to drop the key entirely.
	const out = normaliseTrustedContact({
		name: 'Bob',
		phone: '',
		email: '   ',
		relationship: '',
	});
	assert.deepStrictEqual(out, { name: 'Bob' });
});

test('normaliseTrustedContact: returns null when name is missing or whitespace-only', () => {
	assert.equal(normaliseTrustedContact({ name: '' }), null);
	assert.equal(normaliseTrustedContact({ name: '   ' }), null);
	// Even when other fields are present — without a name there's
	// nothing to display in the notification template.
	assert.equal(
		normaliseTrustedContact({ name: '', phone: '+1 555 0123' }),
		null,
	);
});

// ── normaliseTrustedContacts (the list-level normaliser) ────

test('normaliseTrustedContacts: null / undefined / empty all return []', () => {
	assert.deepStrictEqual(normaliseTrustedContacts(null), []);
	assert.deepStrictEqual(normaliseTrustedContacts(undefined), []);
	assert.deepStrictEqual(normaliseTrustedContacts([]), []);
});

test('normaliseTrustedContacts: filters invalid entries from a mixed list', () => {
	const out = normaliseTrustedContacts([
		{ name: 'Alice', phone: '+1' },
		{ name: '', phone: '+1' }, // dropped — no name
		{ name: '  ' }, // dropped — whitespace-only name
		{ name: 'Bob' },
	]);
	assert.deepStrictEqual(out, [{ name: 'Alice', phone: '+1' }, { name: 'Bob' }]);
});

test('normaliseTrustedContacts: caps at MAX_TRUSTED_CONTACTS', () => {
	// Pin the cap so a runaway editor can't blow up the universal-
	// prefs bag. The cap is intentional product policy, not a
	// technical limit — surface it to the UI rather than failing
	// silently if the user tries to add a sixth.
	const input: TrustedContact[] = Array.from(
		{ length: MAX_TRUSTED_CONTACTS + 3 },
		(_, i) => ({ name: `Contact ${i}` }),
	);
	const out = normaliseTrustedContacts(input);
	assert.equal(out.length, MAX_TRUSTED_CONTACTS);
});

// ── hasReachableChannel ────────────────────────────────────

test('hasReachableChannel: true when phone OR email present', () => {
	assert.equal(hasReachableChannel({ name: 'A', phone: '+1' }), true);
	assert.equal(hasReachableChannel({ name: 'A', email: 'a@b.c' }), true);
	assert.equal(
		hasReachableChannel({ name: 'A', phone: '+1', email: 'a@b.c' }),
		true,
	);
});

test('hasReachableChannel: false when neither phone nor email present', () => {
	// Surfaces the UI warning ("we have no way to reach this person")
	// without blocking the save — the runner may still call manually.
	assert.equal(hasReachableChannel({ name: 'A' }), false);
	assert.equal(hasReachableChannel({ name: 'A', relationship: 'parent' }), false);
});

File: apps/mobile_ios/test/trusted_contacts_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/trusted_contacts.dart';

void main() {
  // Persona-hunt Round 3 finding Woman #4. Dart twin of
  // `apps/web/src/lib/trusted_contacts.test.ts` — keep in lockstep.

  group('normaliseTrustedContact', () {
    test('keeps trimmed name + populated optionals', () {
      final out = normaliseTrustedContact(const TrustedContact(
        name: '  Alice  ',
        phone: ' +44 7700 900111 ',
        email: '  alice@example.com  ',
        relationship: ' partner ',
      ));
      expect(out, isNotNull);
      expect(out!.name, 'Alice');
      expect(out.phone, '+44 7700 900111');
      expect(out.email, 'alice@example.com');
      expect(out.relationship, 'partner');
    });

    test('drops empty-string optionals (does not store "")', () {
      final out = normaliseTrustedContact(const TrustedContact(
        name: 'Bob',
        phone: '',
        email: '   ',
        relationship: '',
      ));
      expect(out, isNotNull);
      expect(out!.name, 'Bob');
      expect(out.phone, isNull);
      expect(out.email, isNull);
      expect(out.relationship, isNull);
    });

    test('returns null when name is missing or whitespace-only', () {
      expect(normaliseTrustedContact(const TrustedContact(name: '')), isNull);
      expect(
          normaliseTrustedContact(const TrustedContact(name: '   ')), isNull);
      expect(
        normaliseTrustedContact(
            const TrustedContact(name: '', phone: '+1 555 0123')),
        isNull,
      );
    });
  });

  group('normaliseTrustedContacts', () {
    test('null / empty returns const []', () {
      expect(normaliseTrustedContacts(null), const <TrustedContact>[]);
      expect(normaliseTrustedContacts(const []), const <TrustedContact>[]);
    });

    test('filters invalid entries from a mixed list', () {
      final out = normaliseTrustedContacts(const [
        TrustedContact(name: 'Alice', phone: '+1'),
        TrustedContact(name: '', phone: '+1'),
        TrustedContact(name: '  '),
        TrustedContact(name: 'Bob'),
      ]);
      expect(out.length, 2);
      expect(out[0].name, 'Alice');
      expect(out[0].phone, '+1');
      expect(out[1].name, 'Bob');
    });

    test('caps at kMaxTrustedContacts', () {
      final input = List.generate(
        kMaxTrustedContacts + 3,
        (i) => TrustedContact(name: 'Contact $i'),
      );
      final out = normaliseTrustedContacts(input);
      expect(out.length, kMaxTrustedContacts);
    });
  });

  group('hasReachableChannel', () {
    test('true when phone OR email present', () {
      expect(
        hasReachableChannel(const TrustedContact(name: 'A', phone: '+1')),
        isTrue,
      );
      expect(
        hasReachableChannel(const TrustedContact(name: 'A', email: 'a@b.c')),
        isTrue,
      );
      expect(
        hasReachableChannel(
            const TrustedContact(name: 'A', phone: '+1', email: 'a@b.c')),
        isTrue,
      );
    });

    test('false when neither phone nor email present', () {
      expect(
        hasReachableChannel(const TrustedContact(name: 'A')),
        isFalse,
      );
      expect(
        hasReachableChannel(
            const TrustedContact(name: 'A', relationship: 'parent')),
        isFalse,
      );
    });
  });

  group('TrustedContact.toJson / fromJson', () {
    test('round-trips with all fields', () {
      const c = TrustedContact(
        name: 'Alice',
        phone: '+1',
        email: 'a@b.c',
        relationship: 'partner',
      );
      final json = c.toJson();
      final back = TrustedContact.fromJson(json);
      expect(back.name, c.name);
      expect(back.phone, c.phone);
      expect(back.email, c.email);
      expect(back.relationship, c.relationship);
    });

    test('omits null optionals from the JSON payload', () {
      const c = TrustedContact(name: 'Bob');
      final json = c.toJson();
      expect(json.containsKey('name'), isTrue);
      expect(json.containsKey('phone'), isFalse);
      expect(json.containsKey('email'), isFalse);
      expect(json.containsKey('relationship'), isFalse);
    });
  });
}
